Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4433
Subject: wayland: Improve interaction of DnD/client cursors after grab
 .
 Fix two issues:
 .
 - In order to follow XDnD protocol, we do let X11 drag sources drive
   pointer cursor updates themselves, between the other responsibilities
   the drag source has in X11. However we were trying to clear the DnD
   cursor at a place that is generic to all DnD operations on a wayland
   compositor (both Xwayland and not).
 - We do this unset of the cursor after detaching the MetaWaylandEventHandler
   and possibly crossing into other handlers (or actors, after a grab). This
   crossing event might have triggered already a change of the cursor by
   itself.
 .
 Both resulted in a staale "default" cursor that the client wouldn't be able
 to undo until the following wl_pointer.enter.
 .
 In order to fix this, make the Xwayland data source check at a more
 generic place that would catch any cursor update attempt from DnD
 code, and shuffle the cursor reset operation so it happens before
 the input handler is detached.
 .
Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/4033
Bug-Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1699942

--- a/src/wayland/meta-wayland-data-device.c
+++ b/src/wayland/meta-wayland-data-device.c
@@ -50,6 +50,7 @@
 #include "wayland/meta-wayland-seat.h"
 #include "wayland/meta-wayland-toplevel-drag.h"
 #include "wayland/meta-wayland-types.h"
+#include "wayland/meta-xwayland-dnd-private.h"
 
 #define ROOTWINDOW_DROP_MIME "application/x-rootwindow-drop"
 
@@ -231,6 +231,12 @@ meta_wayland_drag_grab_set_cursor (MetaW
   g_autoptr (MetaCursorSprite) cursor_sprite = NULL;
   MetaCursorRenderer *cursor_renderer;
 
+#ifdef HAVE_X11_CLIENT
+  /* X11 DnD lets the drag source client drive pointer cursor updates */
+  if (META_IS_WAYLAND_DATA_SOURCE_XWAYLAND (drag_grab->drag_data_source))
+    return;
+#endif
+
   cursor_sprite =
     META_CURSOR_SPRITE (meta_cursor_sprite_xcursor_new (cursor, cursor_tracker));
   cursor_renderer =
@@ -446,6 +452,8 @@ data_device_end_drag_grab (MetaWaylandDr
   MetaDisplay *display = display_from_data_device (data_device);
   MetaCompositor *compositor = meta_display_get_compositor (display);
 
+  meta_wayland_drag_grab_set_cursor (drag_grab, META_CURSOR_DEFAULT);
+
   meta_wayland_drag_grab_set_source (drag_grab, NULL);
   meta_wayland_drag_grab_set_focus (drag_grab, NULL);
 
@@ -478,7 +486,6 @@ data_device_end_drag_grab (MetaWaylandDr
       drag_grab->handler = NULL;
     }
 
-  meta_wayland_drag_grab_set_cursor (drag_grab, META_CURSOR_DEFAULT);
   meta_dnd_wayland_handle_end_modal (compositor);
 
   g_free (drag_grab);
